source/connection.c

```diff
@@ -438,39 +513,60 @@ struct aws_http_server *aws_http_server_new(const struct aws_http_server_options
 
     return server;
 
-error:
-    aws_http_server_destroy(server);
+socket_error:
+    aws_hash_table_clean_up(&server->synced_data.channel_to_connection_map);
+hash_table_error:
+    aws_mutex_clean_up(&server->synced_data.lock);
+mutex_error:
+    aws_mem_release(server->alloc, server);
     return NULL;
 }
 
-void aws_http_server_destroy(struct aws_http_server *server) {
+void aws_http_server_release(struct aws_http_server *server) {
     if (!server) {
         return;
     }
-
     if (server->socket) {
-        /* TODO: Server shutdown probably needs multiple steps, and to be async, like:
-         * 1) stop accepting new connections.
-         * 2) issue shutdown to all existing connections.
-         * 3) wait for connections to finish shutting down.
-         * 4) issue a destroy_complete callback.
-         * 5) finally actually delete server.
-         *
-         * But for now, just assert that there are no outstanding connections */
-        AWS_FATAL_ASSERT(aws_hash_table_get_entry_count(&server->channel_to_connection_map) == 0);
+        bool already_shutting_down = false;
+        /* BEGIN CRITICAL SECTION */
+        s_server_lock_synced_data(server);
+        if (server->synced_data.is_shutting_down) {
+            already_shutting_down = true;
+        } else {
+            server->synced_data.is_shutting_down = true;
+        }
+        s_server_unlock_synced_data(server);
+        /* END CRITICAL SECTION */
 
+        if (already_shutting_down) {
+            /* nothing to do */
+            AWS_LOGF_TRACE(AWS_LS_HTTP_SERVER, "id=%p: The server is already shutting down", (void *)server);
```

Comment:
it might be good to add more, "Why should I care about this log message" info here.

"The service is already shutting down, not doing x"